### ISSUE TITLE
How to handle high peaks in GluonTS simple Feed Forward or DeepAR

### ISSUE THREAD

**Issue by adarwade1** [awslabs/gluon-ts#1112]
(Please state your question in the title, and add here any relevant details,
like a description of the problem, code snippets, examples, and so on.)

**Comment by kaijennissen** [comment]
@adarwade1  What exactly do you mean by peaks? It sound like you are talking about outliers, i.e. a few observations which are far away from the usual level. In this case thera are two options:

- Handle outliers by using the DeepAR model with a robust distribution (StudentT - which is the default).
- Remove outliers during preprocessing.

**Comment by adarwade1** [comment]
Please refer below details.
-Objective : Predict high peaks
Currently we are working for one of the postal services customer. There are multiple post offices and further each post offices having multiple sorting processes to sort different kind of letters. 

In case of one category of letters, the volume will be in chunk on one of the day (third or fourth week of each month). Here requirement is instead of ignoring such peaks (chunk size ) , model should forecast the same high peaks.

We used "week of month" as one of the feature however that also not helping to predict values closer to peak. 

Can you please advise on how we can able to predict high peaks correctly using GluonTS

**Comment by adarwade1** [comment]
Please let me know if you need any further details. GluonTS Simple Feed Forward model is able to forecast other values with better accuracy however there is big difference in actual vs forecasted values in case of peaks.

**Comment by kaijennissen** [comment]
@adarwade1 My first guess is, that you need additional features. From your description I suggest features like weekdays, public holidays and other features which are informative about the peaks.

**Comment by lostella** [comment]
@adarwade1 is `DeepAREstimator` able to understand the presence of peaks in any way (even if not to a great accuracy)? Currently the `SimpleFeedForwardEstimator` does not use any date-related feature, so if those are in any way informative of peaks, maybe adding them would be beneficial

**Comment by adarwade1** [comment]
We used most of fate features to handle seasionality and trends. 
-day of month
-week of month
-week of year
-is weekend
-is holidays
-is non production days. 


Here main challenge is peaks range of value is too high as compare to mean.

**Comment by adarwade1** [comment]
The output is used for resources planning. The day on which more volume requires more resources. That's the reason we need to forecast peaks with better accuracy.

**Comment by kaijennissen** [comment]
@adarwade1 The open question is whether it is possible to forecast the peaks (follow some learnable pattern) or not (peaks are random). As @lostella pointed out, it would be helpful to know if the `DeepAREstimator` is able to detect the peaks and struggles just with the magnitude of the peaks or if peaks are ignored.
Could you share a plot of one example?

**Comment by adarwade1** [comment]
Sure. Let me train DeepAR and share you summary

**Comment by adarwade1** [comment]
Hi All,

Apology for delay in response as busy with some deliverable. Please refer below sample data


![IMG_20201030_192218](https://user-images.githubusercontent.com/73542481/97718564-16ad6b00-1aec-11eb-8142-eab1713e4b96.jpg)
![IMG_20201030_192236](https://user-images.githubusercontent.com/73542481/97718716-452b4600-1aec-11eb-918f-3e5c0fef5d90.jpg)
![IMG_20201030_192247](https://user-images.githubusercontent.com/73542481/97718905-7f94e300-1aec-11eb-9eed-1b59bf07e1c3.jpg)
![IMG_20201030_192256](https://user-images.githubusercontent.com/73542481/97719029-a2bf9280-1aec-11eb-929b-e613161bb94d.jpg)
![IMG_20201030_195105](https://user-images.githubusercontent.com/73542481/97719183-cb478c80-1aec-11eb-8819-fd49fa16225b.jpg)

**Comment by adarwade1** [comment]
The last one is the DeepAR results. Magnitude of peaks are not handle by both DeepAR and SFF.

@kaijennissen  /@lostella , Can you will be able to join Teams meeting if comfortable based on your availability.

**Comment by kaijennissen** [comment]
@adarwade1 Plots look like I expected. DeepARE/SFF do pick up the weekly seasonality but there is currently no feature which explains the peaks. I think your problem to correctly predict the timing of these peaks. Maybe the methods described in this [paper](https://arxiv.org/pdf/1911.10416.pdf) and the corresponding [code](https://github.com/manujosephv/deeprenewalprocess) are helpful.

**Comment by adarwade1** [comment]
Many Thanks for guidance. We will check on update you on the same.

**Comment by adarwade1** [comment]
We used below DeepAR hyperparameters. 
The NegativeBinomialOutput distribution throws nan errors in the first or second epoch. The data doesn't contains nan values, also tried reducing learning rate however not help with NBO distribution. 

It work fine for default distribution

![IMG_20201102_092153.jpg](https://user-images.githubusercontent.com/73542481/97828632-ad954580-1ced-11eb-81ba-d08d7ce7e6df.jpg)

**Comment by adarwade1** [comment]
> @adarwade1 Plots look like I expected. DeepAR/SFF do pick up the weekly seasonality but there is currently no feature which explains the peaks. I think your problem is to correctly predict the timing of these peaks. Maybe the methods described in this [paper](https://arxiv.org/pdf/1911.10416.pdf) and the corresponding [code](https://github.com/manujosephv/deeprenewalprocess) are helpful. 


Ans: These high magnitude peaks present in the 3/4 week of months. So we introduce month of week feature. SFF shown sone improvement in accuracy with inclusion of this feature.

**Comment by adarwade1** [comment]
> @adarwade1 is `DeepAREstimator` able to understand the presence of peaks in any way (even if not to a great accuracy)? Currently the `SimpleFeedForwardEstimator` does not use any date-related feature, so if those are in any way informative of peaks, maybe adding them would be beneficial as @kaijennissen suggests


Can you please share any documents/links which brief on recommendations on DeepAR hyperparameters/ best guidelines if handy. 

We gone thru so many articles however not able to locate any detail doc.

**Comment by lostella** [comment]
I see that some time series in your plots exhibit this spiky behaviour, while others do not. @adarwade1 one thing to try is then to train an estimator only on data that has spikes, and verify that the model can learn that behaviour: that should be the case, if the spikes follow some predictable pattern.

**Comment by adarwade1** [comment]
@lostella /@kaijennissen , 

We train model with seed however everytime we observed different accuracy. 

We used below seed values mentioned in the documents. 
mx.random.seed(0)
np.random.seed(0) 

Can you please advice on the same. 

Thanks and Regards, 
Abhijit

**Comment by lostella** [comment]
@adarwade1 in #1040 I posted a snippet with which I'm getting consistently the same exact results every time, if I'm not wrong. Could you try that?

Also, to avoid confusion, feel free to open a separate issue about this, given that it is unrelated to the original post here.

**Comment by adarwade1** [comment]
@lostella, 
Handling of high peaks , please let us know your availability so that we can connect. 

Thanks and Regards, 
Abhijit

**Comment by adarwade1** [comment]
@lostella,
We tried our best however DeepAR is not able to handle high peaks...Can you please suggest any hyperparameters that can help

**Comment by kaijennissen** [comment]
@adarwade1 Have you considered the case that this is not a problem of the DeepAR Model but of your specific problem?
